libsolidity/experimental/analysis/TypeInference.cpp

```diff
@@ -0,0 +1,1211 @@
+
+	std::optional<Type> underlyingType;
+
+	if (isBuiltIn)
+		// TODO: This special case should eventually become user-definable.
+		underlyingType = m_wordType;
+	else if (_typeDefinition.typeExpression())
+	{
+		ScopedSaveAndRestore expressionContext{m_expressionContext, ExpressionContext::Type};
+		_typeDefinition.typeExpression()->accept(*this);
+
+		underlyingType = annotation(*_typeDefinition.typeExpression()).type;
+	}
+
+	TypeConstructor constructor = typeConstructor(&_typeDefinition);
+	auto [members, newlyInserted] = annotation().members.emplace(constructor, std::map<std::string, TypeMember>{});
+	solAssert(newlyInserted, fmt::format("Members of type '{}' are already defined.", m_typeSystem.constructorInfo(constructor).name));
+	if (underlyingType)
+	{
+		members->second.emplace("abs", TypeMember{helper.functionType(*underlyingType, definedType)});
+		members->second.emplace("rep", TypeMember{helper.functionType(definedType, *underlyingType)});
+	}
+
+	if (helper.isPrimitiveType(definedType, PrimitiveType::Pair))
+	{
+		solAssert(isBuiltIn);
+		solAssert(arguments.size() == 2);
+		members->second.emplace("first", TypeMember{helper.functionType(definedType, arguments[0])});
+		members->second.emplace("second", TypeMember{helper.functionType(definedType, arguments[1])});
+	}
+
+	return false;
+}
+
+bool TypeInference::visit(FunctionCall const&) { return true; }
+void TypeInference::endVisit(FunctionCall const& _functionCall)
+{
+	auto& functionCallAnnotation = annotation(_functionCall);
+	solAssert(!functionCallAnnotation.type);
+
+	Type functionType = type(_functionCall.expression());
+
+	TypeSystemHelpers helper{m_typeSystem};
+	std::vector<Type> argTypes;
+	for (auto arg: _functionCall.arguments())
+	{
+		switch (m_expressionContext)
+		{
+		case ExpressionContext::Term:
+		case ExpressionContext::Type:
+			argTypes.emplace_back(type(*arg));
+			break;
+		case ExpressionContext::Sort:
+			m_errorReporter.typeError(9173_error, _functionCall.location(), "Function call in sort context.");
+			functionCallAnnotation.type = m_typeSystem.freshTypeVariable({});
+			break;
+		}
+	}
+
+	switch (m_expressionContext)
+	{
+	case ExpressionContext::Term:
+	{
+		Type argTuple = helper.tupleType(argTypes);
+		Type resultType = m_typeSystem.freshTypeVariable({});
+		Type genericFunctionType = helper.functionType(argTuple, resultType);
+		unify(functionType, genericFunctionType, _functionCall.location());
+		functionCallAnnotation.type = resultType;
+		break;
+	}
+	case ExpressionContext::Type:
+	{
+		Type argTuple = helper.tupleType(argTypes);
+		Type resultType = m_typeSystem.freshTypeVariable({});
+		Type genericFunctionType = helper.typeFunctionType(argTuple, resultType);
+		unify(functionType, genericFunctionType, _functionCall.location());
+		functionCallAnnotation.type = resultType;
+		break;
+	}
+	case ExpressionContext::Sort:
+		solAssert(false);
+	}
+}
+
+// TODO: clean up rational parsing
```

Comment:
Yeah, all this is half-dead and duplicated code that won't stay like this at all, but well, works for now...